plugins/meta/vrf/main.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2020 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+// VrfNetConf represents the firewall configuration.
+type VrfNetConf struct {
+	types.NetConf
+
+	// Vrf is the name of the vrf to add the interface to.
+	VRFName string `json:"vrfname"`
+}
+
+func main() {
+	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.PluginSupports("0.4.0"), bv.BuildString("vrf"))
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, result, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("missing prevResult from earlier plugin")
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		v, err := findVRF(conf.VRFName)
+
+		if _, ok := err.(netlink.LinkNotFoundError); ok {
+			v, err = createVRF(conf.VRFName)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		err = addInterface(v, args.IfName)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("cmdAdd failed: %v", err)
+	}
+
+	if result == nil {
+		result = &current.Result{}
+	}
+
+	return types.PrintResult(result, conf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	conf, _, err := parseConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		v, err := findVRF(conf.VRFName)
+		if err != nil {
+			return err
+		}
+
+		interfaces, err := assignedInterfaces(v)
+		if err != nil {
+			return err
+		}
+
```

Comment:
So there is no removing of interfaces from VRF because we are expecting that (happy path) a next in chain plugin will destroy the interface, what will in fact also remove it from VRF?

IMO this should also cover unhappy path, so the interface will be left to next in chain plugin for del in the shape in which next plugin left it, with master interface set to what was before.

Reply:
Ah! You are right.
The only thing I am not sure of, is the part where I should restore the master, as the CNI has no memory.

@jellonek what if I unset the master, and let the ADD command return an error if the given interface already has one master set?

Reply:
Yeah, this would be what i would expect. Good idea!